NetworkChat

amzhulanov opened the pull request.

amzhulanov committed:
Создан Клиент и Сервер чата.
Клиент отправляет сообщения серверу, сервер отвечает.

amzhulanov committed:
усложнил задачу.
Теперь сервер всегда ожидает подключения. Подключает несколько клиентов и позволяет выбрать клиента, чтобы отправить ему сообщение. Клиент может разорвать соединение и подключиться заново.

amzhulanov committed:
Добавил комментарии и исключения(catch) с понятным описанием.

amzhulanov committed:
Сообщения отправляются выбранным участникам чата

amzhulanov committed:
Сообщения отправляются выбранным участникам чата

amzhulanov committed:
Сообщения отправляются выбранным участникам чата

amzhulanov committed:
Пользователь выделяется из списка и  ему можно отправить сообщение

amzhulanov committed:
Заработало удаление пользователя из списка активных пользователей при закрытии окна

amzhulanov committed:
Сделана авторизация через БД MS SQL

amzhulanov committed:
Добавил возможность регистрации пользователей.

amzhulanov committed:
Поправил закрытие соединения с БД

amzhulanov committed:
мелкие правки

amzhulanov committed:
Доделал нормальную регистрацию

amzhulanov committed:
Merge pull request #9 from amzhulanov/NetworkChat-test1

Доделал нормальную регистрацию

amzhulanov committed:
Настроена регистрация через userRepository

amzhulanov committed:
Перенёс объявление переменной userRepository в метод main
Убрал Static

usharik commented on src/Java/ru/geekbrains/NetworkChat/Authorization/AuthServiceJdbcImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+public class AuthServiceJdbcImpl implements AuthService {
+
+    private static UserRepository userRepository;
```
Почему static? Из классов Service и Repository ни один не должен быть static.

usharik commented on src/Java/ru/geekbrains/NetworkChat/Authorization/AuthServiceJdbcImpl.java:
```diff
@@ -0,0 +1,43 @@
+    public void registrationUser(User user)  {
+        String login = user.getLogin();
+        String password=user.getPassword();
+        userRepository.insert(login,password);
```
Почему бы не передавать целый класс User в метод insert

amzhulanov replied:
Исключительно ради наглядности. Или это влияет на время исполнения?
Поправил на передачу User

amzhulanov committed:
Убрал Static у поля userRepository

amzhulanov committed:
В метод insert() теперь передаётся User, вместо отдельных логин/пароль

amzhulanov committed:
Для каждого клиента создаётся свой файл с историей сообщений.
Сообщения записываются при отправке и получении.
При открытии чата подгружаются сообщения из истории.
Осталось ограничить загрузку сообщений из истории в кол-ве 100 строк.

amzhulanov committed:
При открытии чата клиенту подгружается история переписки, ограничена 100 строками.
Для этого я считываю весь файл в MAP и из МАП вывожу последние 100 строк.

p.s. много времени пришлось потратить, пока настроил парсинг строки, при выводе из файла в окно чата. Особенно с форматом даты промучился.

amzhulanov committed:
Доработал, чтобы в окне чата строки шли по дате от старых к новым
определяю с какой строки выводить сообщение.
Например, если всего 115 строк, то вывод пойдёт с 15-ой строки

usharik commented on src/Java/ru/geekbrains/NetworkChat/Client/HistoryMessage.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    public void writeMessage(LocalDateTime created, String login,String userFrom,String textMessage) throws IOException {
```
В качестве параметра было бы лучше передавать экземпляр класса Message

amzhulanov replied:
Если передавать экземпляр класса, то тогда придётся отдельно передавать имя файла, в который записывать сообщение.

Вызов этого метода происходит в двух случаях:
1. клиент отправляет - тут запись сообщения происходит в файл отправителя
2. клиент слушает - тут запись сообщения происходит в файл получателя

В этот метод я передаю значения:
1. Время сообщения
2. имя файла, в который будет записываться сообщение
3. от кого поступило сообщение
4. текст сообщения.

amzhulanov replied:
А, вспомнил) в вашей версии кода вы на сервере меняете местами отправителя и получателя.
В этом случае для записи в файл можно передать класс Message, вы правы.
Но я решил, что менять местами отправителя и получателя не правильно. Т.к. я смотрю с точки зрения клиента отправителя.

usharik commented on src/Java/ru/geekbrains/NetworkChat/Client/HistoryMessage.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    public void readMessage(String login,MessageReciever messageReciever) throws IOException {
```
Не очень хорошая идея передавать в этот класс ссылку на MessageReciever. Он является сервисом только для сохранения и считывания сообщений. Совместным использованием этих сервисов будет заниматься основной класс клиента чата.

amzhulanov replied:
MessageReciever выбрал, т.к. в нём есть метод submitMessage. Его использую для вывода сообщения из файла в чат пользователя.
Попробую исправить

amzhulanov committed:
Внёс правки, учитывая разбор ДЗ на уроке
